lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,256 @@
+pub struct SshBackend {
+    program: OsString,
+    allowed_signers: Option<OsString>,
+}
+
+#[derive(Debug, Error)]
+pub enum SshError {
+    #[error("SSH sign failed with exit status {exit_status}:\n{stderr}")]
+    Command {
+        exit_status: ExitStatus,
+        stderr: String,
+    },
+    #[error("Failed to run ssh-keygen: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Signing key required")]
+    MissingKey {},
+    #[error("Allowed signers file not provided")]
+    MissingAllowedSigners {},
+}
+
+impl From<SshError> for SignError {
+    fn from(e: SshError) -> Self {
+        SignError::Backend(Box::new(e))
+    }
+}
+
+impl SshBackend {
+    pub fn new(program: OsString, allowed_signers: Option<OsString>) -> Self {
+        Self {
+            program,
+            allowed_signers,
+        }
+    }
+
+    pub fn from_config(config: &config::Config) -> Self {
+        Self::new(
+            config
+                .get_string("signing.backends.ssh.program")
+                .unwrap_or_else(|_| "ssh-keygen".into())
+                .into(),
+            config
+                .get_string("signing.backends.ssh.allowed-signers")
+                .map_or_else(|_| None, |value| Some(value.into())),
+        )
+    }
+
+    fn run(&self, input: &[u8], args: &[&OsStr]) -> Result<Vec<u8>, SshError> {
+        let process = Command::new(&self.program)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .args(args)
+            .arg("-n")
+            .arg("git")
+            .spawn()?;
+
+        process.stdin.as_ref().unwrap().write_all(input)?;
+        let output = process.wait_with_output()?;
+
+        if !output.status.success() {
+            Err(SshError::Command {
+                exit_status: output.status,
+                stderr: String::from_utf8_lossy(&output.stderr).into(),
+            })
+        } else {
+            Ok(output.stdout)
+        }
+    }
+
+    fn get_allowed_signers(&self) -> Result<OsString, SshError> {
+        if let Some(allowed_signers) = &self.allowed_signers {
+            Ok(allowed_signers.into())
+        } else {
+            Err(SshError::MissingAllowedSigners {})
+        }
+    }
+
+    fn find_principal(&self, signature_file_path: &OsStr) -> Result<Option<String>, SshError> {
+        let allowed_signers = self.get_allowed_signers()?;
+
+        let process = Command::new(&self.program)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .args([
+                "-Y".as_ref(),
+                "find-principals".as_ref(),
+                "-f".as_ref(),
+                allowed_signers.as_os_str(),
+                "-s".as_ref(),
+                signature_file_path,
+            ])
+            .spawn()?;
+
+        let output = process.wait_with_output()?;
+
+        let result: String = String::from_utf8_lossy(&output.stdout).into();
```

Comment:
Is data loss allowed here?

I don't know what will be returned. If it's supposed to be ASCII for example, maybe better to error out if it isn't.

Reply:
Ok I reworked this to expect utf8 or error out.